Docs 5945 insight topic limitations hn (#265)
mulesoft/docs-runtime-manager: modules/ROOT/pages/insight.adoc

endif::[]
:keywords: cloudhub, analytics, monitoring, insight, filter

image:logo-cloud-active.png[link="/runtime-manager/deployment-strategies", title="CloudHub"]
image:logo-hybrid-active.png[link="/runtime-manager/deployment-strategies", title="Hybrid Deployment"]
image:logo-server-disabled.png[link="/runtime-manager/deployment-strategies", title="Anypoint Platform Private Cloud Edition"]
image:logo-pcf-disabled.png[link="/runtime-manager/deployment-strategies", title="Pivotal Cloud Foundry"]
image:logo-cloud-active.png[link="/runtime-manager/deployment-strategies#cloudhub", title="CloudHub"]
image:logo-hybrid-active.png[link="/runtime-manager/deployment-strategies#hybrid", title="Hybrid"]
image:logo-server-disabled.png[link="/runtime-manager/deployment-strategies#anypoint-platform-private-cloud-edition", title="Anypoint Platform PCE"]
image:logo-rtf-disabled.png[link="/runtime-manager/deployment-strategies#anypoint-runtime-fabric", title="Runtime Fabric"]

[NOTE]
====
Insight is not supported on Anypoint Platform Private Cloud Edition.
====

> Insight is a troubleshooting tool that gives you in-depth visibility into business transactions and events on your Mule apps deployed through Runtime Manager. Insight makes information searchable and helps you find and recover from any errors that occurred during message processing by tracking everything your data does in an app and replaying your transactions instantly.

Reviewer: Edit instructions:

I have a PR for you to review. Rewrite/refresh of a topic + some new content. It says 17 files, but it's really just the one (insight.adoc). Sometime tomorrow? Thanks!

> For Mule 3 apps only, Insight supports the ability to replay transactions for advanced error recovery. Replay is not supported for Mule 4 apps.

Reviewer: Do we need this restatement?

Author: Yes.

For this reason, do not enable Insight for long periods for production environments.
Use Insight for troubleshooting only.

> For information on tools to use for monitoring a production environment, see xref:monitoring.adoc[Monitoring Applications and Servers].

Reviewer: information on > information about
GLOBAL

* What was the result of a transaction?
* If something went wrong during processing, at what point did the failure occur?

After you enable Insight, the following events are tracked by default:

> * Data that passes through endpoints (inbound and outbound) in your Mule applications

Reviewer: Not worded as an event, per list intro.
"Data passing through endpoints..."?

* Flow initiation and completion
* Custom business events that you embed in your Mule flows

== Limits and Limitations  

The following limits apply to events:

* A single application is limited to 50,000 events per minute.
* All applications in a single business group are limited to 100,000 events per minute. 

Events over these limits are dropped. 

Also, no more than 20 custom event definitions per application definition are stored.  

Insight does not support visibility into batch processing. See xref:mule-runtime::batch-processing.adoc[Batch Processing].

> Insight does not preserve replay data when switching regions. Ensure that this doesn't violate any of your compliance requirements.

Reviewer: Do we need to remind audience that this applies only to M3?

> Logs for applications deployed to on-premises servers contain event information, but not message contents.

Reviewer: ```suggestion
Logs for applications deployed to on-premises servers contain event information but not message contents.
```

=== Data Persistence

Insight does not limit the size of transactions. Insight displays all transactions in your application for 30 days and keeps events for two days.

> Replay data for applications created before October 7, 2017 is stored in the default region, us-east-1.

Reviewer: ```suggestion
Replay data for applications created before October 7, 2017, is stored in the default region us-east-1.
```

> Data for applications deployed, or deployed to a different region, after October 7, 2017 is stored in the region where the application is deployed.

Reviewer: I did not understand the difference between deployed and to a different region at first, so I think this probably needs rewording:
"Data for applications (initially?) deployed or (redeployed? moved?) to a different region after October 7, 2017, is stored in the region where the application is (most recently?) deployed.

=== Payload Transformation

> The following MIME types are always mapped as Strings. 

Reviewer: ```suggestion
The following MIME types are always mapped as Strings:
```

[source,linenums]

> ----

Reviewer: Should these be in alpha order if no other?

application/javascript
application/json
application/xml
application/csv
application/rtf
application/vnd.mozilla.xul+xml
application/x-csh
application/x-sh
application/xml+html
application/xml
images/svg+xml
text/calendar
text/css
text/csv
text/html
text/xml
text/plain
----

> For all other types, Mule Agent attempts to transform payloads to another format, based on the Stream type that Mule runtime engine exposes to the Mule Agent.

Reviewer: Mule Agent?
I haven't seen this term yet.
When I look it up, I see first references to Anypoint Runtime Manager agent and, later, in blog posts, Mule agent.

Reviewer: I just got weigh-in from PMM: Mule agent

> For example, Mule Agent converts payloads in `CURSOR_STREAM_PROVIDER` type to a byte array.

Reviewer: Do we need "type" if we format the name as syntax?

> Any Streams that are not mapped to String by default appear in the logs. 

Reviewer: Should this be...:
```suggestion
Any streams that are not mapped to String by default appear in the logs. 
```

== Enable or Disable Insight

To enable Insight to track the events your Mule servers generate:

. Sign in to your Anypoint Platform account.
. Click *Runtime Manager*.
. Click *Applications* in the menu on the left, and click the application in the *File* column to open the management panel on the right side of the page. 
. In the management panel, click *Insight*.
. Click *Settings* in the menu on the left, and then click the *Insight* tab:
+
.The arrow shows the *Insight* tab on the *Settings* page.
image::rtm-insight-tab.png[Insight tab in the Settings page]
. Select one of the tracking options:
+
--
** *Disabled* to disable Insight
** *Metadata* to track only events

> ** *Metadata and Replay* to track events and enable the ability to instantly replay a transaction from the console.

Reviewer: no period
```suggestion
** *Metadata and Replay* to track events and enable the ability to instantly replay a transaction from the console
```

+
This option is available only for Mule 3 applications deployed to CloudHub.
Replay only works for flows that have inbound endpoints.
+

> For more information, see <<Diagnosing Problems and Replaying Transactions>>.

Reviewer: Is this a placeholder for a xref?

Author: No, it's an intra-topic xref.

--

> . Select the checkbox to confirm that you want to enable Insight and click *Apply Changes*.

Reviewer: ```suggestion
. Select the checkbox to confirm that you want to enable Insight, and click *Apply Changes*.
```

. Click *Apply Changes* to save the application settings.

== Searching and Filtering Transactions and Events

The Insight dashboard displays information at three tracking levels: 

* *Transactions*
+

> Logical groupings of related events that often correspond to a business view of the system.

Reviewer: Not a sentence; no period:
```suggestion
Logical groupings of related events that often correspond to a business view of the system
```

* *Events*
+

> Low-level details of a transaction.

Reviewer: ditto
```suggestion
Low-level details of a transaction
```

+
Events map to message processors and endpoints and reveal information about exceptions and any custom business events you might have configured.
* *Metadata*
+

> Customized key-value pairs that you specify as part of custom events in your application.

Reviewer: ```suggestion
Customized key-value pairs that you specify as part of custom events in your application
```

+
The detail that Insight tracks at this level provides information about high-level business events.

The following image shows where to find transactions, events, and metadata on the Insight dashboard:

.The screenshot shows (*1*) the Transactions level, (*2*), the Events level, and (*3*), the Metadata level on the Insight dashboard.
image::rtm-insight-dashboard.png[Tracking levels on the Insight dashboard]

> Business data, if configured, appear in the Metadata area (*3*) of the dashboard. See <<Setting Up Business Events>>.

Reviewer: ```suggestion
Business data, if configured, appears in the Metadata area (*3*) of the dashboard. See <<Setting Up Business Events>>.
```
Is << >> OK?

=== Search or Filter Transactions

On the *Insight* page, you can search or filter data to refine the results displayed. You can search for:

* A specific transaction by transaction ID
* Transactions within a specific date range

> * Transactions according to flow name, exception message, processing time, and/or any custom business data

Reviewer: Avoid "and/or" and use instead the most likely or common for the context. GLOBAL

+

> To enable the *Business Data* field your app must have business events configured. See <<Setting Up Business Events>>.

Reviewer: I'll stop asking about << >>

To search or filter transactions:

. Click *Insight* in the menu on the left.
. Enter your search criteria in the *Search* field.
. Click *Advanced* in the *Search* field.
. Specify filter options:
+
.The screenshot shows the advanced search filter fields on the Insight dashboard.
image::rtm-insight-search-filter.png[Advanced search filter fields on the Insight dashboard]
// image::chintransid.png[CHInTransID]
+
--

> * To display only failed or completed transactions, select *Failed* or *Completed* from the menu in the *Transaction ID* field.

Reviewer: colon
```suggestion
* To display only failed or completed transactions, select *Failed* or *Completed* from the menu in the *Transaction ID* field:
```

+
.The arrow shows the menu in the *Transaction ID* field.
image::rtm-insight-filter-transactions.png[Advanced search filter fields on the Insight dashboard]

* To display transactions within a specific date range, specify a date or range in the *Date & Time* fields.
+

> Alternatively, click and drag within the graph to select a specific date range:

Reviewer: ```suggestion
Alternatively, drag within the graph to select a specific date range:
```

+
.The arrow shows the selected transaction range on the Insight dashboard.
image::rtm-insight-filter-selected.png[Selected transaction range on the Insight dashboard]
--
. Click *Apply* on the *Advanced* pane to apply your search criteria.

> . If you want to save a search, click *Save search*:

Reviewer: I don't think that the actions performed on a saved search are really subordinate tasks to saving a search: running, editing, and deleting.

+
.The arrow shows the *Save search* icon on the Insight dashboard.
image::rtm-insight-save-search.png[Save search icon on the Insight dashboard]
+
Runtime Manager saves the search to the name you specify.

* To run a saved search, click the search name.

> * To edit a saved search, hover the mouse over the search name, and then click the *Edit* link.

Reviewer: ```suggestion
* To edit a saved search, hover over the search name and click the *Edit* link:
```

To display the filter options, click the *Transaction details* menu for a transaction:

.The arrow shows the *Transaction details* menu.
image::rtm-insight-trans-details.png[Transaction details menu]

The default settings of the events filter depend upon your user role:

* *Support*
+
The default events filter setting displays only custom events.
* *Admin* or *Developer*
+
The default events filter setting displays custom events, endpoint events, and message processor events, but does not display flow events.

== Diagnosing Problems and Replaying Transactions

When Insight reports a failed transaction, you can diagnose the problem and fix the root issue. 

For Mule 3 applications and flows with inbound endpoints, you can use the Insight message replay feature to replay the transaction that failed.
 
=== Transaction Failed Status

For example, when a transaction fails because an application exceeds the limited number of API calls for a Salesforce account, Insight displays the status of the transaction as `Failed`. 

> Click to expand the transaction, then examine the individual events within that transaction. The exception and the error message are listed next to the failed event, as well as an existing stack trace:

Reviewer: ```suggestion
Click to expand the transaction, and then examine the individual events within that transaction. The exception and the error message are listed next to the failed event, as well as an existing stack trace:
```

image::chexcmsg.png[CHExcMsg]

> === Replay Transaction (Mule 3 apps only)

Reviewer: I suggest putting the parenthetical qualification into the text directly following the title rather than in it.

After investigating the logs and resolving the issue (for example, by purchasing more API capacity) that caused the transaction failure, you can replay the transaction.

> Replay is not supported for Mule 4 apps.

Reviewer: Nec?

> . Click the replay icon (circular arrow) to the right of the first event in the failed transaction to replay it:

Reviewer: I just saw Valkyrie use a screen cap for an icon; suggest the same here.

Reviewer: Also, if we show it in the capture directly following, we might not need to describe it here.

+
.The screenshot shows (*1*) the replay icon and (*2*) the *Purge Replay Data* button.
image::rtm-insight-replay-purge.png[Replay icon and Purge Replay Data button]
. At the prompt, click *Replay Transaction* to confirm.
+
Insight displays an information popup indicating that the transaction was replayed and displays the replayed transactions immediately on the dashboard and in the *Logs* tab.

> === Purge Replay Transaction Data

Reviewer: This head does not use qualifier, so why use it earlier? They are peers after all.

To delete all existing replay data for that application, click the *Purge Replay Data* button.

[WARNING]
You can't restore replay data that has been purged.

> == Insight Status States 

Reviewer: Table of Insight Status States?
No intro text to set a context for the table?

[%header,cols="20,60a"]
|===
|Status |Description

> |`FAILED`|The event failed while being processed and has an associated exception associated.

Reviewer: ```suggestion
|`FAILED`|The event failed while being processed and has an associated exception.
```

|`COMPLETED`|The event processed fully and successfully.
|`IN_PROGRESS`|The event has started and has not yet completed.

> |`STALLED`|The event has been put on hold due to the maximum number of events to be processed per transaction (10000). Any event above this mark is set as stalled and then processed after this event is no longer stalled.

Reviewer: ```suggestion
|`STALLED`|The event has been put on hold due to exceeding the maximum number of events to be processed per transaction (10000). Any event above this number is set as stalled and then processed after the number of events to be processed is no longer exceeded.
```
